### ISSUE TITLE
Make code more Pythonic

### ISSUE THREAD

**Issue by RichardBronosky** [arpruss/raspberryjammod#6]
I'd like to make the python scripts in https://github.com/arpruss/raspberryjammod/tree/master/python2-scripts/mcpipy and https://github.com/arpruss/raspberryjammod/tree/master/python3-scripts/mcpipy comply with PEP8.

I'd also like to discuss what can be done to improve scripts like https://github.com/arpruss/raspberryjammod/blob/master/python3-scripts/mcpipy/castle.py#L5 (specifically line 5) and https://github.com/arpruss/raspberryjammod/blob/master/python3-scripts/mcpipy/mc.py#L7 to teach good coding habits like avoiding `import *`.

What do you think of this? I know I could do it in my fork, but I'd rather not split our resources and further confuse the young coders (and parents) about which project they should use.

I really think that RaspberryJamMod is the best way to get people started with programming Minecraft via Python. I have purchased my daughter 2 books on the subject. I believe you are familiar with [Adventures in Minecraft](http://www.wiley.com/WileyCDA/WileyTitle/productCd-111894691X.html) which suggests using a Bukkit server (though Martin O'Hanlon now offers the option of using a CanaryMod server). Just a few days ago [Learn to Program with Minecraft](https://www.nostarch.com/programwithminecraft) was released. It instructs readers to run a Spigot server. I have tried all 4 methods and I truly think that your solution of using Forge is the best. However, I think that it's very important the all the Python code in this project be exemplary.

To be clear, I think your work is very impressive, even amazing. It's just that it has the feel of my personal projects where I get things working and don't go back to make sure I've done things Pythonicly. When I do work professionally, or to open source, I go through and have a clean-up phase where I try to consider the needs/conventions of the community. I think with a little work RaspberryJamMod + Forge could and should be the basis for every Minecraft programming book, course, and summer camp (which I was shocked to discover not only exists but there were half a dozen different ones being taught in my area alone).

Would you be open to engaging in this with me?

**Comment by arpruss** [comment]
I do the import * stuff on purpose: I don't want middle-school kids who haven't done any coding to have to worry about the details of what gets imported from the module: it would just make the start of learning too steep. I don't want to have to initially explain classes or leave things like block.GLASS with a mysterious "block.". In the summer course I was teaching. Students make enough typos that even so a fair amount of the time I was just going from computer to computer correcting a bit of indentation here, a missing import there, etc. 

I want them to be able to do things like:
  from mcturtle import *
  t = Turtle()
  t.penblock(GLASS)
  for i in range(10):
    t.go(10)
    t.right(36)

Yes, certainly, for more advanced students one should keep the namespaces intact. But what I have seems a good compromise as it stands.

**Comment by RichardBronosky** [comment]
I'm glad to hear that you are actively using this to teach children. I am really torn about what is "best" in our unique shared context. I do love the simplicity of `from mc import *` and struggled with giving it up. This is the kind of discussion I'd like us to have.

However with a decent IDE (i've got my daughter using Atom) the autocompletion benefit of using block.GLASS is very compelling. Otherwise, she had to navigate to the source file to see what blocks are available. It's far more intuitive to have them grouped under "block." Is it really that mysterious when all the turtle commands above are grouped under "t." or others are under an "mc." namespace?

I don't mean to argue. I just want to discuss the options.

What book do you use to teach your students?

**Comment by arpruss** [comment]
On reflection, having <code>import mcpi.block as block</code> instead of <code>from block import *</code> in mc.py would have been a better choice for the reason you give. Nonetheless, autocomplete (at least in Python Tools for Visual Studio) will pick up the blocks in the global namespace, except that you have to have some idea what the names of the blocks are. The cluttered global namespace does make it harder to use the debugger, though.

But I don't want to change mc.py in a backwards-incompatible way. A lot of people have seen my instructables for RaspberryJamMod, and I would be breaking any code they have.

The best bet would be to make a replacement for mc.py. But I don't know what to call it. The natural option is minecraft.py, but there already is an mcpi/minecraft.py, and some scripts by other authors call for that to be present in the main directory. And I want the name to be brief and self-explanatory. Suggestions are welcome. 

Maybe there is no need. After all, mc.py is just a convenience module. You can always do what they do in the O'Hanlon-Whale book, just importing the mcpi modules manually.

I used the O'Hanlon-Whale book for some ideas. And got some more ideas from O'Hanlon's discussion board at Stuff about Code.

**Comment by RichardBronosky** [comment]
All valid points. I'll take some time to think about them before I respond.

I notice that you mention Visual Studio. The 2 books I got my daughter each use IDLE. After 2 days of trying to stick with the book to keep it simple, I found it to be unbearable. It certainly took some configuring to get it optimal for Python and Minecraft, but I find GitHub's Atom editor to be excellent. Have you considered trying to move to an IDE that didn't require Windows? If so I can walk you through everything I did on my Mac and on her Windows 10 PC to get it working smoothly.

Here are some of the points/features/concepts:
1. She does her work in a Dropbox folder that is shared with me so I can quickly review her work and help her with problems.
  - (This could be awesome for you and your students!)
  - Inside the shared folder you create symbolic links to `~\\AppData\\Roaming\\.minecraft\\mcpipy` (What? [Windows has symlinks](http://superuser.com/questions/1020821/how-to-create-a-symbolic-link-on-windows-10)?
1. Her code is validated in realtime thanks to https://atom.io/packages/linter-flake8
1. She has awesome autocompletion thanks to https://atom.io/packages/autocomplete-python
1. When she's editing, she can run her script within Atom with a simple (configurable) keystroke thanks to https://atom.io/packages/script
1. When she gets a script to where she likes it she just drags a copy into her "Mick Pip EE" folder.
  - She thinks it's awesome that she can create commands like `/python trefoilhack` on her own.
  - (I got her started in all this just be turning her loose with your sample scripts and she thought it was amazing all the cool things she could do with your code. Then I showed her that **I could make simple changes** to your code to do things like make fancytree cyan. Of course she wanted to be able to do it. Then she quickly wanted to be able to make more significant changes. That's when I got her started with the books.)
1. I have already introduced her to good version control habits which was easy once I bought her a pack of cute octocat stickers to decorate her computer https://twitter.com/RichardBronosky/status/682115824115658752
1. Committing and pushing code is easy with https://atom.io/packages/git-plus

As you can see from those links to Atom packages, the community is thriving!
1. Hundreds of thousands of downloads for significant user contributed packages
2. Everything is on github so if you want to modify one or see examples on how to create your own...
1. It just became a cultural norm to use an animated gif to demo and that makes "plugin shopping" super fast.

Just something to think about.

**Comment by arpruss** [comment]
Thanks for all the suggestions. I haven't used Atom yet. Does Atom have a good interactive debugger for python? PTVS has an excellent python debugger. I didn't introduce the kids to the debugger, so it was only useful to me, but it was nice to use the same tools as the kids do.

**Comment by RichardBronosky** [comment]
I hope you don't get annoyed by all these suggestions. It's just really apparent to me (by all the parents that approach us at coffee shops while I'm teaching her and ask me if I could teach their kids) that there is a serious need to have the discussion of how teach this. I feel bad doing it in O'Hanlon's board because he has finished an awesome book and I don't want to tear it apart after the fact. It is in-fact the reason I am even getting to consider these things.

I haven't looked into debugger packages yet. I only started using Atom for my daughter's sake. I use VIM for my daily work and I have a whole system for debugging with either bpython, ipython, or pdb depending on the case. I'll look into the debugger because my current solution for shell exploration still uses IDLE but launches with mc.py preloaded. I'd rather have her be able to "run code until current line" as she would call it (aka: breakpoint)

**Comment by arpruss** [comment]
I've made one switch in how the namespaces work: I replaced mc.py and mcturtle.py with mine.py and mineturtle.py in my demo scripts. The one difference is that if you do "from mine import *", block constants are all in the block module rather than globals. Thus, block.AIR instead of AIR. So the global namespace is much less cluttered now. This not only makes code a bit more pythonic, but it also makes it easier to use a debugger or auto-complete.
